infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java

```diff
@@ -257,116 +336,167 @@ public static void main(String[] args) throws Exception {
+     * Corfu Server initialization.
+     *
+     * @param serverContext Initialized Server Context.
+     */
+    public CorfuServer(@Nonnull ServerContext serverContext) {
+        this(serverContext,
+                ImmutableMap.<Class, AbstractServer>builder()
+                        .put(BaseServer.class, new BaseServer(serverContext))
+                        .put(SequencerServer.class, new SequencerServer(serverContext))
+                        .put(LayoutServer.class, new LayoutServer(serverContext))
+                        .put(LogUnitServer.class, new LogUnitServer(serverContext))
+                        .put(ManagementServer.class, new ManagementServer(serverContext))
+                        .build()
+        );
     }
 
     /**
-     * Setup logback logger
-     *  - pick the correct logging level before outputting error messages
-     *  - add serverEndpoint information
+     * Corfu Server initialization.
      *
-     * @param opts command line parameters
-     * @throws JoranException logback exception
+     * @param serverContext Initialized Server Context.
+     * @param serverMap     Server Map with all components.
      */
-    private static void configureLogger(Map<String, Object> opts) {
-        final Logger root = (Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
-        final Level level = Level.toLevel(((String) opts.get("--log-level")).toUpperCase());
-        root.setLevel(level);
+    public CorfuServer(@Nonnull ServerContext serverContext,
+                       @Nonnull Map<Class, AbstractServer> serverMap) {
+        this.serverContext = serverContext;
+        this.serverContext.setBindToAllInterfaces(bindToAllInterfaces);
+        this.serverMap = serverMap;
+        router = new NettyServerRouter(new ArrayList<>(serverMap.values()));
+        this.serverContext.setServerRouter(router);
     }
 
     /**
-     * Creates all the components of the Corfu Server.
-     * Starts the Server Router.
+     * Start the Corfu Server by listening on the specified port.
      *
-     * @param opts Options passed by the user.
+     * @return Listening Channel Future.
      */
-    public static void startServer(Map<String, Object> opts, boolean bindToAllInterfaces) {
+    public ChannelFuture start() {
+        bindFuture = startAndListen(serverContext.getBossGroup(),
+                serverContext.getWorkerGroup(),
+                CorfuServer::configureBootstrapOptions,
+                serverContext,
+                router,
+                (String) serverContext.getServerConfig().get("--address"),
+                Integer.parseInt((String) serverContext.getServerConfig().get("<port>")));
+
+        return bindFuture.syncUninterruptibly();
+    }
 
-        int port = Integer.parseInt((String) opts.get("<port>"));
+    @Override
+    public synchronized void close() {
+        if (!close) {
+            log.info("close: Shutting down Corfu server and cleaning resources");
+            close = true;
```

Comment:
Can you please use AtomicBoolean. It is always much cleaner and less fragile than using volatile

Reply:
Not required. The close is synchronized and the variable is used only in close.

Reply:
Are you going to replace volatile by atomic when you decide to update your variable? Don't you find that this approach is very fragile? 
Also, you must be very careful and check that all method which use volatile variable must be synchronized.
Also, it's error-prone and requires anyone else who uses this volatile variable to be careful when they decide to bring new logic.

So what the point for using volatile? Make code more dangerous and fragile but justify this by the following statement: I update the volatile variable only in synchronized methods, then it should be ok to use volatile

Reply:
I don't think AtomicBoolean is the same as volatile memory-wise and unless we use compareAndSet semantics provided by AtomicBoolean, I don't find it useful.
But in this case, I have replaced synchronized with compareAndSet.